### PR TITLE
Allow for PKCE flow without client secret + add docs

### LINKED ISSUE #17107: OAuth PKCE authentication requires divulging client secret
While Gitea supports PKCE as an OAuth provider, it doesn't allow clients to authenticate without providing a client secret. Clients which cannot safely store a client secret, e.g. serverless single-page apps and mobile apps, thus cannot authenticate. In particular, Netlify's authentication flow is done entirely client-side, and its OAuth PKCE authentication flow thus will not work with Gitea currently.

My current proposal for fixing this is to add a configuration option for OAuth applications which determines whether their clients are [public or confidential](https://oauth.net/2/client-types/). Public clients will be required to use PKCE but are not required to provide a client secret, whereas confidential clients will be required to provide a client secret but are not required to use PKCE.

### LINKED ISSUE #25047: OAuth should not require secret for public clients
According to OAuth spec https://datatracker.ietf.org/doc/html/rfc8252#section-8.5

> it is NOT RECOMMENDED for authorization servers to require client authentication of public native apps clients using a shared secret

but Gitea requires client secret in both authorization code flow and refresh flow regardless of client type

https://github.com/go-gitea/gitea/blob/7a5873335afb00dda5fa123fb023c88278e79deb/routers/web/auth/oauth.go#L756

https://github.com/go-gitea/gitea/blob/7a5873335afb00dda5fa123fb023c88278e79deb/routers/web/auth/oauth.go#L698

Fix would be to validate secret if and only if app.ConfidentialClient

### PR DESCRIPTION
**Pull request by denyskon**
The PKCE flow according to [RFC 7636](https://datatracker.ietf.org/doc/html/rfc7636) allows for secure authorization without the requirement to provide a client secret for the OAuth app.

It is implemented in Gitea since #5378 (v1.8.0), however without being able to omit client secret.
Since #21316 Gitea supports setting client type at OAuth app registration.

As public clients are already forced to use PKCE since #21316, in this PR the client secret check is being skipped if a public client is detected. As Gitea seems to implement PKCE authorization correctly according to the spec, this would allow for PKCE flow without providing a client secret.

Also add some docs for it, please check language as I'm not a native English speaker.

Closes #17107
Closes #25047

### CODE REVIEW

**Comment by 6543** [comment]
Do you have a client i can test against in recommendation?

**Comment by denyskon** [comment]
> Do you have a client i can test against in recommendation?

Not really.... you can use one of the web-based oauth tools like https://oauthdebugger.com/ or the cli, but I do not know what else could be used

File: docs/content/doc/development/oauth2-provider.en-us.md [hunk 1]
```diff
@@ -134,7 +136,69 @@ For public clients, a redirect URI of a loopback IP address such as `http://127.
...
+   The user will now be asked to authorize your application. If they authorize it, the user will be redirected to the `REDIRECT_URL`, for example:
+
+   ```curl
+   https://[REDIRECT_URI]?code=RETURNED_CODE&state=STATE
+   ```
+
+2. Using the provided `code` from the redirect, you can request a new application and refresh token. The access token endpoint accepts POST requests with `application/json` and `application/x-www-form-urlencoded` body, for example:
+
+   ```curl
+   POST https://[YOUR-GITEA-URL]/login/oauth/access_token
```
Comment (techknowlogick) [inline docs/content/doc/development/oauth2-provider.en-us.md:179]: Would this endpoint need CORS headers to be set, as I assume w/ Decap it is client side HTTP requests being made to fetch the access_token?

Comment (denyskon) [reply docs/content/doc/development/oauth2-provider.en-us.md:179]: Yes, CORS is required for it to work. But Decap is entirely based on client-side http requests, so  as far as I can remember CORS headers were needed for every used request type.

File: routers/web/auth/oauth.go
Comment (hickford) [inline routers/web/auth/oauth.go:1]: There are two calls to ValidateClientSecret in this file. Surely need to make this change for both?

Comment (denyskon) [reply routers/web/auth/oauth.go:1]: You're completely right, I didn't notice it. Added confidential client check for refresh token.

**Comment by denyskon** [comment]
Added a test for PKCE with public client.

**Review by 6543 — APPROVED** [review]

**Review by wxiaoguang — APPROVED** [review]

**Pull request merged**